### PR TITLE
System contracts

### PR DESCRIPTION
**Pull request by rowandh**
System contracts, like smart contracts, can conceptually be thought of a function `f(x)` which when applied to a state returns a new state, `state' = f(state)`. The state is stored in a trie and its root is committed to a block header.

System contracts have the following differences to smart contracts:
* Are instantiated from types and not by reflecting bytecode
* Deployed via the codebase, not on-chain
* Don't consume gas
* Don't store receipts (currently)
* Ignore any value attached to a call transaction (currently)
* Will typically require some sort of authentication to invoke
* Can be used to extend the protocol

To achieve dynamic dispatch required by allowing contracts to be invoked on-chain, I've added the concept of a dispatcher. The dispatcher is responsible for instantiating the contract, and mapping the on-chain method call and parameters to the instance. It's registered with the DI container and can use any dependency. Each on-chain system contract must implement a dispatcher, but contracts that don't need to be called on-chain are just regular types and don't need them.

Non-system contracts achieve dynamic dispatch using reflection, which places limits on things like method parameter types and return types. There will be a limited number of system contracts, so the additional overhead of implementing a dispatcher for each one is not too great.

I added two example contracts, `AuthContract` and `DataStorageContract` to demonstrate how contract interactions can work.

There's currently two currently unused classes (see `Rules` and `Mining`) that I wrote to get an understanding of a few things. Eg. if we decide we don't need receipts or value, we don't need a coinview rule and can use a regular full validation consensus rule.

Future work:
* Change serialization of the call data if necessary
* Decide if receipts are necessary
* Refund unused value attached to a transaction

### CODE REVIEW

**Comment by quantumagi** [comment]
@rowandh , why introduce differences between smart contracts and system contracts when there is no stated requirement to do 
so? 

The only difference should be:

- System contracts are still "smart contracts" with the only real difference being that they can **only be deployed by us**. 
  - We are adding the code-based deployment to "smart contracts", to be enabled on STRAX only (for now).
    - This was already achieved by removing the standard whitelisting mechanism and by limiting ourselves
      to deployment via the codebase (for now).
    - The option of allowing deployment from the code-base is something we could support on the Cirrus side as well and 
      gain the same benefit of enhanced functionality that we allow for contracts deployed in this way.

Let me respond to each seemingly ad-hoc difference that you are suggesting:

> Are instantiated from types and not by reflecting bytecode
A contract is still a class/type. .NET always executes bytecode.

> Deployed via the codebase, not on-chain
As mentioned that should have been just a deployment option - i.e. not a difference per se. The same could be an option for "normal" smart contracts. I am using quoted here as there really should be no differentiation.
> Don't consume gas
This only needs to apply to read-only calls initiated from the codebase. I.e. no on-chain tx involved.

> - Don't store receipts (currently)
> - Ignore any value attached to a call transaction (currently)
We should still be updating the same code base so why have we lost this ability?
  
>  Will typically require some sort of authentication to invoke
True, but the same authentication mechanism should be usable for "normal" smart contracts so not really a "difference".

> Can be used to extend the protocol
That could be true for both Strax and Cirrus if simply by virtue of a different deployment option.

**Comment by rowandh** [comment]
@quantumagi there's really not that much different. Did you actually look at the code yet?

**Comment by quantumagi** [comment]
> @quantumagi there's really not that much different. Did you actually look at the code yet?

@rowandh , so far I've only responded to the summary.  I will have a look at the code tomorrow.

**Comment by rowandh** [comment]
Alright, because I don't understand the aggressive nitpick of the summary.

**Comment by quantumagi** [comment]
> Alright, because I don't understand the aggressive nitpick of the summary.

Don't take it the wrong way. It's just an honest response to what is written.

**Comment by quantumagi** [comment]
An option that we have discussed adding is to store contract state on a per-contract rather than a per-address basis as it appears useful for code-based contracts. Noting, perhaps for future PRs, that that may mean sharing the balance among the different addresses of the same contract. I.e. we would be able to deploy a new version of a contract (with a different address) and still have access to the full balance via the new as well as the old addresses. We have to consider if this impacts the wallet.

**Comment by rowandh** [comment]
> we don't need smart contracts AND system contracts - rather we need options within smart contracts.

Adding new options to the smart contracts feature to make it do both smart contracts AND system contracts things violates composability - we'd be modifying something existing rather than composing something new.

As for code reuse - the existing SC interfaces and some existing implementations have been used, which is why this plugs nicely into the existing coinview rule. Everything is composed with the concrete interface implementations in the new feature.

Consider what would happen if we had modified the existing SC feature to use the system contract concrete types. As-is, it would break the Cirrus smart contract implementation. We could fix it by doing something dirty, like adding some kind of option to the smart contracts feature which replaces and removes the unnecessary services, but why would we do that when we have OOP 🙂.

If you look at `Stratis.Features.SystemContracts` module it's pretty simple:
- Namespace containing the system contract-specific types
- New feature class, which is just a way to compose the DI container with only the dependencies we need. Most of them are the same as smart contracts, ones unnecessary for system contracts have been removed.

**Comment by rowandh** [comment]
> We have previously discussed the option of storing contract state on a per-contract rather than a per-address basis

Any system contract can masquerade as any other contract and access its balance, state etc. To support a version upgrade we'd just need to add way to know what instance to instantiate based upon block height/BIP9s. This isn't done in this PR but should be relatively trivial to add.

**Comment by quantumagi** [comment]
> Consider what would happen if we had modified the existing SC feature to use the system contract concrete types. As-is, it would break the Cirrus smart contract implementation.

System contracts should not be portrayed as solely using concrete types when there is not even such a stated requirement for system contracts. Also, smart contracts already use concrete types under the covers (e.g. via `ISmartContractState`) and as such system contracts could do likewise. On "interfacing" level (public methods) only normal smart contract types need to be used.

**Comment by rowandh** [comment]
> We since added the "embeddability" option which (we hope) could also make them more "powerful.

Right and to do so the main difference is that we don't reflect new instances from compiled bytecode that's stored in a database. We just instantiate new instances like any other object. And because we can do that, and we because don't have on-chain trust concerns, we don't need to force contracts to inherit `SmartContract` or be constrained in their dependencies.

There's almost nothing else different about this PR.

**Comment by quantumagi** [comment]
> Right and to do so the main difference is that we don't reflect new instances from compiled bytecode that's stored in a database. We just instantiate new instances like any other object.

Correct.

> And because we can do that, and we because don't have on-chain trust concerns, 

Correct.

> we don't need to force contracts to inherit SmartContract or be constrained in their dependencies.

Not inheriting from `SmartContract` can be both good and bad. If this is what makes smart contracts sufficiently different from "system contracts", by your definition, to warrant a separate feature then maybe consider leaving this as a future upgrade or separate mechanism.

**Comment by rowandh** [comment]
It's not necessary to make system contracts need to inherit from `SmartContract`. `SmartContract` is a paradigm put in place to help with writing contracts deployed in a decentralized trustless environment with limited dependencies. When that's not a concern it's actually extra work to support it.

It would still be possible to do so. I've published an additional system contract example which uses `SmartContract` to demonstrate how it would work.

File: src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs [hunk 1]
```diff
@@ -116,6 +121,15 @@ public Result<object> Dispatch(ISystemContractTransactionContext context)
 
                         return Result.Fail<object>($"Method {context.CallData.MethodName} overload with {context.CallData.Parameters.Length} params does not exist on type {nameof(DataStorageContract)} v{context.CallData.Version}");
 
+                    case nameof(DataStorageContract.PersistComplexType):
+                        var txRawHex = context.CallData.Parameters[0] as string;
+
+                        var tx = Transaction.Parse(txRawHex, RawFormat.Satoshi);
+
+                        instance.PersistComplexType(tx);
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs:129]: Let's just pass complex types the same way we would for smart contracts - i.e. by serializing them to byte arrays. There is no requirement to support complex types on public methods.

**Comment by quantumagi** [comment]
> To achieve dynamic dispatch required by allowing contracts to be invoked on-chain, I've added the concept of a dispatcher. The dispatcher is responsible for instantiating the contract, and mapping the on-chain method call and parameters to the instance.

Let's avoid the additional complexity of a dispatcher as we don't have any requirement to process complex types (over public methods) other than the types we already support for smart contracts.

File: src/NBitcoin/Network.cs [hunk 2]
```diff
@@ -186,7 +186,7 @@ public class PrimaryAuthenticators
 
         public uint Quorum { get; private set; }
 
-        public PrimaryAuthenticators(Network network, string[] base58Addresses, uint quorum)
+        public PrimaryAuthenticators(string[] base58Addresses, uint quorum)
```
Comment (quantumagi) [inline src/NBitcoin/Network.cs:189]: This is removing an unused argument and can be moved to a separate PR to improve the reviewability of this PR.

File: src/Stratis.Bitcoin.Networks/StraxRegTest.cs [hunk 3]
```diff
@@ -86,7 +86,7 @@ public StraxRegTest()
                 new Dictionary<KeyId, string> { },
                 new Dictionary<uint256, (int start, int? end)[]> { },
                 new Dictionary<uint256, (string, bool)> { },
-                new PrimaryAuthenticators(this, new[]
+                new PrimaryAuthenticators(new[]
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Networks/StraxRegTest.cs:89]: https://github.com/stratisproject/StratisFullNode/pull/536/files#r623626624

File: src/Stratis.Bitcoin.Networks/StraxTest.cs [hunk 4]
```diff
@@ -83,7 +83,7 @@ public StraxTest()
                 new Dictionary<KeyId, string> { },
                 new Dictionary<uint256, (int start, int? end)[]> { },
                 new Dictionary<uint256, (string, bool)> { },
-                new PrimaryAuthenticators(this, new[]
+                new PrimaryAuthenticators(new[]
```
Comment (quantumagi) [inline src/Stratis.Bitcoin.Networks/StraxTest.cs:86]: https://github.com/stratisproject/StratisFullNode/pull/536/files#r623626624

File: src/Stratis.SmartContracts.Networks/SmartContractsPoSRegTest.cs [hunk 5]
```diff
@@ -80,7 +80,7 @@ public SmartContractsPoSRegTest()
                 new Dictionary<KeyId, string> { },
                 new Dictionary<uint256, (int start, int? end)[]> { },
                 new Dictionary<uint256, (string, bool)> { },
-                new PrimaryAuthenticators(this, new[]
+                new PrimaryAuthenticators(new[]
```
Comment (quantumagi) [inline src/Stratis.SmartContracts.Networks/SmartContractsPoSRegTest.cs:83]: https://github.com/stratisproject/StratisFullNode/pull/536/files#r623626624

File: src/Stratis.StraxD/Properties/launchSettings.json [hunk 6]
```diff
@@ -1,7 +1,8 @@
 {
   "profiles": {
     "Stratis.StraxD": {
-      "commandName": "Project"
+      "commandName": "Project",
```
Comment (quantumagi) [inline src/Stratis.StraxD/Properties/launchSettings.json:4]: Did you mean to change the launch settings?

File: src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs (hunk 1 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs:129]: This is just an example, when implementing it you can choose never to do this.

File: src/Stratis.Features.SystemContracts/Identifier.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Features.SystemContracts
+{
+    /// <summary>
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/Identifier.cs:7]: Embedded contract's `version + contract id` is equivalent in purpose to the `transaction hash` associated with on-chain deployment. Based on this insight a `pseudo-hash` (`uint256`) was introduced together with our BIP 9 integration. Consider using that instead.

**Comment by rowandh** [comment]
The dispatcher pattern removes significant complexity that is present with the method dispatch using reflection that the other SC approach uses. It's not necessary here because we control the types that are deployed and don't need to support arbitrary user-generated types. It's more performant for us and clearer to use the dispatcher pattern.

File: src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+using Stratis.SmartContracts.Core.State;
+
+namespace Stratis.Features.SystemContracts.Contracts
+{
+    /// <summary>
+    /// Sample contract that uses auth and stores data in the state.
+    /// </summary>
+    public class DataStorageContract
+    {
+        public DataStorageContract(IStateRepositoryRoot state, Network network, AuthContract auth)
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/Contracts/DataStorageContract.cs:15]: How does this contract specify the version of the AuthContract?

**Comment by quantumagi** [comment]
System Contracts = Contracts that are only deployed by us. The dictionary contract (includes white-listing) plus the embedded contract container would list all system contracts. System Contracts don't have to be embedded contracts.
Embedded Contracts = Contracts embedded into the code-base and implicitly self-deploy with the code when a given BIP is reached.

File: src/Stratis.StraxD/Program.cs [hunk 9]
```diff
@@ -52,11 +52,7 @@ public static async Task Main(string[] args)
                     .UseSmartContractPosPowMining()
                     .UseApi()
                     .AddRPC()
-                    .AddSmartContracts(options =>
-                    {
-                        options.UseReflectionExecutor();
-                        options.UsePoSWhitelistedContracts();
-                    })
+                    .AddSystemContracts()
```
Comment (quantumagi) [inline src/Stratis.StraxD/Program.cs:55]: This is probably not the only daemon that requires updating. For instance some fedpeg daemons would be creating mainchain daemons when called with the -mainchain argument.

File: src/Stratis.Features.SystemContracts/Identifier.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    /// Defines the identity of a system contract. Allows the identity scheme to be flexible.
+    /// Using a ValueType is desirable here because it be used as a dictionary key.
+    /// </summary>
+    public struct Identifier
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/Identifier.cs:11]: Just use a `uint160` or `EmbeddedContractIdentifier`. Both would include the version so you won't need to pass that separately.

**Comment by quantumagi** [comment]
Please merge latest changes plus what you can take from PR 534 - especially the latest tweaks to the embedded contract container and surrounding classes.

**Comment by quantumagi** [comment]
I don't think we should forego the ability to account for the amount of gas consumed by embedded contracts. This may still matter when this feature is ported to Cirrus - especially for scenarios where "normal" contracts may be calling embedded contracts.

File: src/Stratis.Features.SystemContracts/DispatcherRegistry.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Features.SystemContracts
+{
+    public interface IDispatcherRegistry
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/DispatcherRegistry.cs:6]: Can the dispatch mechanism be implemented in such a way that any smart contract can "opt into" providing a dispatch implementation rather than making it mandatory?

File: src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace Stratis.Features.SystemContracts.Tests
+{
+    public class EndToEndTests
+    {
+        [Fact]
+        public void Complete_Execution_Success()
+        {
+            var network = new StraxMain();
+            var authContract = new AuthContract.Dispatcher(network.SystemContractContainer);
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs:18]: Can we increase the similarity between embedded smart contracts and on-chain smart contracts by making the same dispatch mechanism available to both - even if in one case the framework may be invoking it? I.e. increase the amount of common code being executed.

File: src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+            return signatures[0] == "secret";
+        }
+
+        /// <summary>
+        /// Dynamic method resolution and dependency injection for the <see cref="AuthContract"/> type
+        /// to allow it to be invoked on-chain.
+        /// 
+        /// Why is this necessary?
+        /// For system contracts, call data is received from a transaction and deserialized to strings.
```
Comment (quantumagi) [inline src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs:68]: This should be true for both embedded contracts and on-chain contracts. We should standardize on a common approach for both. I.e. "normal" smart contracts should also be able to use dispatchers so that we reach a common code path sooner.

**Comment by quantumagi** [comment]
@rowandh , can we get more eyes on this PR?

File: src/Stratis.Features.SystemContracts/Identifier.cs (hunk 10 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/Identifier.cs:11]: EmbeddedContractIdentifier didn't even exist when I wrote this

File: src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs (hunk 12 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs:18]: We could but that's a substantial change to smart contracts and outside the scope of this work.

File: src/Stratis.Features.SystemContracts/DispatcherRegistry.cs (hunk 11 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/DispatcherRegistry.cs:6]: We could but that's a substantial change to smart contracts and outside the scope of this work.

File: src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs (hunk 13 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs:68]: We could but that's a substantial change to smart contracts and outside the scope of this work.

**Comment by rowandh** [comment]
> @rowandh , there is a lot going on in this PR. To make it easier to review please split this PR along these boundaries:
> 
> * a PR that implements the embedded contract capability (regardless of network) (for PoS/PoA),
> * a PR that implements the dispatcher in conjunction with overridable method of the smart contract base class (regardless of network).  "Normal" smart contracts would be using the default (non-overridden) implementation. and
> * a PR that implements system contracts on Strax. I am in favour of using the established options based approach - e.g. adding or omitting `options.AllowOnChainDeployment()` but I will let others weigh in on that as well. The only real-difference for system contracts would be the suppression of on-chain deployment. All other capabilities can be shared by both.

I can split this up but not on these lines, because:
* We don't need embedded contract capability yet
* We don't need dispatchers for contracts yet
* We don't need the smart contract execution model

Instead I can do these PRs:
* Add system contract implementations of `IContractExecutor`
* Implement system contracts on Strax

**Comment by quantumagi** [comment]
> We don't need embedded contract capability yet
Not sure I understand. Are you suggesting removing the ability to embed contract code from PR 536?

> We don't need dispatchers for contracts yet
Same here. Are you planning on removing the dispatcher code currently present in PR 536?

> We don't need the smart contract execution model
Can you highlight the areas we don't need? In was thinking we could keep overall gas plumbing but remove the implicit gas accounting/observing and the accompanying code rewriting/caching requirement.

File: src/Stratis.Features.SystemContracts/DispatcherRegistry.cs (hunk 11 above)
Comment (quantumagi) [reply src/Stratis.Features.SystemContracts/DispatcherRegistry.cs:6]: On the contrary, I am asking that we don't make changes to the current smart contracts in the regards mentioned.

File: src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs (hunk 12 above)
Comment (quantumagi) [reply src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs:18]: It has the same scope as this more substantial changes with a large dispatcher implementation as well as mandatory dispatching code being added to the contracts themselves.

File: src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs (hunk 13 above)
Comment (quantumagi) [reply src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs:68]: On the contrary, the idea is to reduce change by being very selective and minimal with what is changed.

**Comment by rowandh** [comment]
> Not sure I understand. Are you suggesting removing the ability to embed contract code from this (or derived) PR(s)?

These are just classes, they don't need to be tied to contracts in any way except to modify the state tree.

> Same here. Are you planning on removing the dispatcher code currently present in this PR?

No, because this does not change anything to do with how contracts are written on Cirrus.

> Can you highlight the areas we don't need? I suggest we keep overall gas plumbing but remove the _implicit_ gas accounting/observing and the accompanying code rewriting/caching requirement. I.e. it would be beneficial to _not_ remove any code that may assist with _explicit_ gas accounting in the future. Can we also preserve any plumbing related to sending funds to the contract or spending funds and pretty much everything else one would expect from a "normal" smart contract. Sounds like a big ask, but actually that should require less change.

We don't need to charge gas for calling these methods. But if we _do_ then we can, likewise with handling funds.

File: src/Stratis.Features.SystemContracts/Identifier.cs (hunk 10 above)
Comment (quantumagi) [reply src/Stratis.Features.SystemContracts/Identifier.cs:11]: Its just the renamed SystemContractIdentifier which did exist.

File: src/Stratis.Features.SystemContracts/DispatcherRegistry.cs (hunk 11 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/DispatcherRegistry.cs:6]: The change you're suggesting in #542 affects how contracts are able to be written on Cirrus and is significant enough that would require updating documentation, further testing etc. It's more changes, not less.

The dispatcher implementation looks large but it's just boilerplate to handle dynamic dispatch and it's easy to understand.

File: src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs (hunk 12 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs:18]: The changes in this PR have no impact on the way contracts are written for Cirrus.

File: src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs (hunk 13 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/Contracts/AuthContract.cs:68]: It does the exact opposite though, by adding new functionality to Cirrus.

File: src/Stratis.Features.SystemContracts/DispatcherRegistry.cs (hunk 11 above)
Comment (rowandh) [reply src/Stratis.Features.SystemContracts/DispatcherRegistry.cs:6]: Another distinction is that Cirrus APIs are public APIs. Other people use them and they have permanence. These APIs are internal and much easier for us to change in the future.

**Comment by quantumagi** [comment]
Please merge the latest `infracontracts` with this PR.

File: src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs (hunk 12 above)
Comment (quantumagi) [reply src/Stratis.Features.SystemContracts.Tests/EndToEndTests.cs:18]: > The changes in this PR have no impact on the way contracts are written for Cirrus.

That may be the goal and I understand the motivation (reducing changes to Smart Contracts) but it does go against code re-usability to not invent a common mechanism. I suppose you can justify it and put that justification up for scrutiny.
